pulser/tests/test_sequence_sampler.py

```diff
@@ -0,0 +1,373 @@
+    for seq in seqs:
+
+        global_keys = [
+            ("Global", basis, qty)
+            for basis in ["ground-rydberg", "digital"]
+            for qty in ["amp", "det", "phase"]
+        ]
+        local_keys = [
+            ("Local", basis, qubit, qty)
+            for basis in ["ground-rydberg", "digital"]
+            for qubit in seq._qids
+            for qty in ["amp", "det", "phase"]
+        ]
+
+        sim = pulser.Simulation(seq)
+        want = sim.samples
+        got = sample(seq)
+
+        for addr, basis, qty in global_keys:
+            try:
+                np.testing.assert_array_equal(
+                    got[addr][basis][qty], want[addr][basis][qty]
+                )
+            except KeyError:
+                np.testing.assert_array_equal(
+                    got[addr][basis][qty],
+                    np.zeros(len(got[addr][basis][qty])),
+                )
+
+        for addr, basis, qubit, qty in local_keys:
+            try:
+                np.testing.assert_array_equal(
+                    got[addr][basis][qubit][qty],
+                    want[addr][basis][qubit][qty],
+                )
+            except KeyError:
+                np.testing.assert_array_equal(
+                    got[addr][basis][qubit][qty],
+                    np.zeros(len(got[addr][basis][qubit][qty])),
+                )
+
+
+def test_inXY() -> None:
+    """Test sequence in XY mode."""
+    pulse = Pulse(
+        BlackmanWaveform(200, np.pi / 2),
+        RampWaveform(200, -np.pi / 2, np.pi / 2),
+        0.0,
+    )
+
+    reg = pulser.Register.from_coordinates(np.array([[0.0, 0.0]]), prefix="q")
+    seq = pulser.Sequence(reg, MockDevice)
+    seq.declare_channel("ch0", "mw_global")
+    seq.add(pulse, "ch0")
+    seq.measure(basis="XY")
+
+    sim = pulser.Simulation(seq)
+
+    want = sim.samples
+    got = sample(seq)
+
+    for qty in ["amp", "det", "phase"]:
+        np.testing.assert_array_equal(
+            got["Global"]["XY"][qty], want["Global"]["XY"][qty]
+        )
+
+    for q in seq._qids:
+        for qty in ["amp", "det", "phase"]:
+            try:
+                np.testing.assert_array_equal(
+                    got["Local"]["XY"][q][qty], want["Local"]["XY"][q][qty]
+                )
+            except KeyError:
+                np.testing.assert_array_equal(
+                    got["Local"]["XY"][q][qty],
+                    np.zeros(len(got["Local"]["XY"][q][qty])),
+                )
+
+
+def test_modulation(mod_seq: pulser.Sequence) -> None:
+    """Test sampling for modulated channels."""
+    got = sample(mod_seq, modulation=True)["Global"]["ground-rydberg"]["amp"]
+
+    chan = mod_seq.declared_channels["ch0"]
+    input = np.pi / 2 / 0.42 * np.blackman(1000)
+    want = chan.modulate(input)
+
+    np.testing.assert_allclose(got, want, atol=1e-2)
+    # Equality at 1e-2 only... Why are they not equal? channel.modulate is
+    # called in both cases.
```

Comment:
Indeed !

(0.42 is the exact value for the Blackman window)

Reply:
The test now passes with `np.testing.assert_array_equal`, very nice 👏

Reply:
> (0.42 is the exact value for the Blackman window)

You're right, my bad. It just differs because the discrete nature of the samples means that the sum of all the values might not be exactly 0.42.